quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/cdi/KeycloakBeanProducer.java

```diff
@@ -17,21 +17,22 @@
 
 package org.keycloak.quarkus.runtime.integration.cdi;
 
-import jakarta.enterprise.context.ApplicationScoped;
-import jakarta.enterprise.context.RequestScoped;
-import jakarta.enterprise.inject.Produces;
-import org.keycloak.common.util.Resteasy;
+import io.quarkus.arc.Unremovable;
+
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory;
 
-import io.quarkus.arc.Unremovable;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.context.RequestScoped;
 
 @ApplicationScoped
 @Unremovable
 public class KeycloakBeanProducer {
 
-    @Produces
     @RequestScoped
     public KeycloakSession getKeycloakSession() {
-        return Resteasy.getContextData(KeycloakSession.class);
+        KeycloakSessionFactory sessionFactory = QuarkusKeycloakSessionFactory.getInstance();
```

Comment:
Even though the session should be built only once it is not clear who is managing it and that is a big problem if there is no proper logic to encapsulate this. Also, if for some reason, the request scope is created twice (don't ask me why/how) we don't need to worry about any internal detail of the underlying stack.

I'm more in favor of having the creation from a single place and where the session lifecycle is managed.

Reply:
I am not exactly sure what kind of information you are looking for, but I will say that RESTEasy Reactive handles the activation / deactivation / reactivation / destruction of the CDI request scope be delegeating to Arc

Reply:
I'm really not sure if creating the session here is the best. Perhaps I'm alone on this but I would like to see what others have to say about it.

Reply:
Probably good to add that the session creation is lazy, so most of the time it will happen via the TransactionSessionHandler. However this change does allow for users who wish to access the KeycloakSession in a prematching request filter to do so.

Reply:
But the pre-matching phase runs in the event loop, doesn't it?

Reply:
It does. That would currently be considered unsupported usage, but it does offer continunity with what was possible before the switch to resteasy reactive.

Reply:
I see. Using a session in a thread other than a worker thread should not be something we want to support, I would say. Just to say that the prematching working is not really an argument in favor of what you are proposing :) I would rather be very hard on not allowing that (or failing to do so) at all.

Reply:
My main point is that CDI producers are really nice but not for this kinda of object for which we need a very clear mechanism for its lifecycle. If we could introduce something that also leverages CDI to destroy this session at the end of the request processing, then I think it would be more consistent.

Reply:
> My main point is that CDI producers are really nice but not for this kinda of object for which we need a very clear mechanism for its lifecycle.

Previously we were still tied to the request scope, just at a later point and via some casting that provided access to the vertx routing context.

> If we could introduce something that also leverages CDI to destroy this session at the end of the request processing, then I think it would be more consistent.

We can, that would replace the registerCompletionCallback callback - it didn't seem necessary to pursue that though.

> Even though not aligned with my initial argument around having the marker interface for components involved in the request/transaction lifecycle.

I'm fine with adding cdi destroy logic, just need to make sure it runs in the right thread. - either way I think you still need the CloseSessionFilter, so it doesn't help simplify things much.